lib/biaser.js

```diff
@@ -0,0 +1,217 @@
+    // pick the least awful combo, we also ensure that we load spread the
+    // failures around.
+    if (killrate === Infinity) {
+      return info.kills;
+    }
+
+    // If we are able to calculate what the bias is, we want to weigh it a
+    // little heavier so that we avoid that combo and
+    let bias = 1 + this.__killRateMultiplier * killrate;
+    return bias;
+  } else {
+    // Like if we have zero spawns and zero kills, we want to give this combo
+    // a slight preference
+    return this.__emptyComboBias;
+  }
+};
+
+/**
+ * Given a set of instance types and regions, fetch the information needed to
+ * determine a bias value
+ */
+Biaser.prototype.fetchBiasInfo = async function(azinfo, types) {
+  let regions = Object.keys(azinfo);
+
+  let biasTable = {};
+
+  let spotKillQuery = [
+    'select count(id) from AwsProvisioner.InstanceTerminated',
+    'where time > now() - ' + this.__statsAge,
+    'and stateChangeCode <> \'Client.UserInitiatedShutdown\'',
+    'and stateChangeCode <> \'Client.InstanceInitiatedShutdown\'',
+    'group by az, region, instanceType',
+  ].join(' ');
+
+  let spotFulfilledQuery = [
+    'select count(id) from AwsProvisioner.SpotRequestsFulfilled',
+    'where time > now() - ' + this.__statsAge,
+    'group by az, region, instanceType',
+  ].join(' ');
+
+  let fullQuery = spotKillQuery + ' ; ' + spotFulfilledQuery;
+  debug(fullQuery);
+
+  let results = await this.influx.query(fullQuery);
+  assert(results);
+  assert(results.length === 2);
+  let kills = results[0];
+  let spawns = results[1];
+
+  // Validate that we got things in the shape we expect
+  assert(kills.name === 'AwsProvisioner.InstanceTerminated');
+  assert(kills.columns[1] === 'count');
+  assert(kills.columns[2] === 'az');
+  assert(kills.columns[3] === 'region');
+  assert(kills.columns[4] === 'instanceType');
+  for (let killpt of kills.points) {
+    let killcount = killpt[1];
+    let zone = killpt[2];
+    let region = killpt[3];
+    let type = killpt[4];
+    if (!biasTable[region]) {
+      biasTable[region] = {};
+    }
+    if (!biasTable[region][zone]) {
+      biasTable[region][zone] = {};
+    }
+    if (!biasTable[region][zone][type]) {
+      biasTable[region][zone][type] = {
+        kills: 0,
+        spawns: 0,
+      };
+    }
+    biasTable[region][zone][type].kills = killcount;
+  }
+
+  // Validate that we got things in the shape we expect
+  assert(spawns.name === 'AwsProvisioner.SpotRequestsFulfilled');
+  assert(spawns.columns[1] === 'count');
+  assert(spawns.columns[2] === 'az');
+  assert(spawns.columns[3] === 'region');
+  assert(spawns.columns[4] === 'instanceType');
+  for (let spawnpt of spawns.points) {
+    let spawncount = spawnpt[1];
+    let zone = spawnpt[2];
+    let region = spawnpt[3];
+    let type = spawnpt[4];
+    if (!biasTable[region]) {
+      biasTable[region] = {};
+    }
+    if (!biasTable[region][zone]) {
+      biasTable[region][zone] = {};
```

Comment:
Good example of something that would be a lot simpler, if you didn't have nested objects...

Using `this.__biasTable[region+ '/' + zone + '/' + type]` is much easier... fewer corner cases...
If you must use nested objects to avoid issues with `/` potentially being used in a region, zone or type name please consider using lodash `_.get` and `_.set`, e.g. `_.set({}, ["a", "b", "c"], 4)`.

IMO, if you're scared that `/` might be used in a region, zone or type name, it might be better to just make a utility function that concatenates these into a string for you and escapes the string names... or something like that.